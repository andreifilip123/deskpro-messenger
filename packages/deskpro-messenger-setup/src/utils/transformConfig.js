export default (settings) => {
  const screens = {
    index: {
      screenType: 'Blocks',
      blocks: []
    }
  };

  if (settings.getIn(['chat', 'enabled'])) {
    const chatBlockConfig = settings.getIn(['messenger', 'chat']);
    screens.index.blocks.push({
      blockType: 'StartChatBlock',
      title: chatBlockConfig.get('title', 'Conversation'),
      description: chatBlockConfig.get('description'),
      linkText: chatBlockConfig.get('buttonText', 'Start a chat'),
      showAgentPhotos: chatBlockConfig.get('showAgentPhotos'),
      to: 'startChat'
    });
    screens.startChat = settings.get('chat').toJS();
    screens.startChat.screenType = 'StartChatScreen';
    delete screens.startChat.enabled;
    if (screens.startChat.noAnswerBehavior === 'save_ticket') {
      screens.startChat.ticketFormConfig = [
        {
          fields: [
            {
              name: 'name',
              label: 'Full Name',
              type: 'text',
              validation: ['required'],
              placeholder: 'John Doe'
            },
            {
              name: 'email',
              label: 'Email',
              type: 'text',
              validation: ['required'],
              placeholder: 'john.doe@company.com'
            }
          ]
        }
      ];
    }
  }

  if (settings.getIn(['tickets', 'enabled'])) {
    const ticketBlockConfig = settings.getIn(['messenger', 'tickets']);
    screens.index.blocks.push({
      blockType: 'ScreenLink',
      blockTitle: ticketBlockConfig.get('title'),
      description: ticketBlockConfig.get('description'),
      label: ticketBlockConfig.get('buttonText'),
      to: 'newTicket'
    });
    screens.newTicket = settings.get('tickets').toJS();
    delete screens.newTicket.enabled;
  }

  const enabledGreetings = [null];
  if ('object' === typeof screens.startChat) {
    enabledGreetings.push('/screens/startChat');
  }
  const config = {
    screens,
    themeVars: {
      '--color-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--color-background': settings.getIn(['styles', 'backgroundColor'], '#3d88f3'),
      '--brand-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--brand-secondary': settings.getIn(['styles', 'backgroundColor'], '#a9b0b0'),
      '--text-color': settings.getIn(['styles', 'textColor'], '#ffffff'),

    },
    greetings: {},
    enabledGreetings
  };

  return config;
};
