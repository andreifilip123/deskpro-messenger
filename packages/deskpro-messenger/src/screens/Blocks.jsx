import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { injectIntl } from 'react-intl';

import Block from '../components/core/Block';
import QuickSearchBlock from '../components/search/QuickSearchBlock';
import Button from '../components/form/Button';
import { connect } from 'react-redux';
import { getAgentsAvailable, canUseChat } from '../modules/info';
import { getActiveChat } from '../modules/chat';
import AvatarHeads from '../components/ui/AvatarHeads';
import ScrollArea from 'react-scrollbar/dist/no-css';
import { Footer } from '../components/core/Footer';

const transMessages = {
  startChatTitle: {
    id: 'blocks.start_chat.title',
    defaultMessage: 'Chat with us'
  },
  startChatDescription: {
    id: 'blocks.start_chat.description',
    defaultMessage: 'Need help? Just reply to start a live conversation with one of our team'
  },
  startChatLink: {
    id: 'blocks.start_chat.link',
    defaultMessage: 'Start Chat'
  },
  continueChatTitle: {
    id: 'blocks.continue_chat.title',
    defaultMessage: 'Continue existing chat'
  },
  continueChatLink: {
    id: 'blocks.continue_chat.link',
    defaultMessage: 'Continue'
  },
  ticketsTitle: {
    id: 'blocks.tickets.title',
    defaultMessage: 'Your Tickets'
  },
  ticketsViewAllLink: {
    id: 'blocks.tickets.view_all_link',
    defaultMessage: 'view all'
  }
};

const blocksMapping = {
  StartChatBlock: injectIntl(({ to, intl, linkText, activeChat, ...props }) => {
    let title = '';
    let link = '';
    let description = '';
    if(activeChat) {
      link = intl.formatMessage(transMessages.continueChatLink, props);
      title = description = intl.formatMessage(transMessages.continueChatTitle, props);
    } else {
      link = intl.formatMessage(linkText ? { id: linkText, defaultMessage: linkText } : transMessages.startChatLink, props);
      title = props.title || intl.formatMessage(transMessages.startChatTitle);
      description = props.description || intl.formatMessage(transMessages.startChatDescription);
    }
    return (
      <Block title={title}>
        {props.showAgentPhotos && !activeChat ? <AvatarHeads agentsAvailable={props.agentsAvailable} /> : null}
        {!activeChat &&
          <div className="dpmsg-BlockText">
            {description}
          </div>
        }
        <Button to={`/screens/${to}`} width="full" color="primary">
          {link}
        </Button>
      </Block>
    )
  }),
  QuickSearchBlock: injectIntl(({ intl }) => (
    <QuickSearchBlock title='Quick Search' />
  )),
  TicketsBlock: injectIntl(({ intl, ...props }) => (
    <Block title={intl.formatMessage(transMessages.ticketsTitle)}>
      <Link title='Browse your tickets' to="/screens/tickets">
        {intl.formatMessage(transMessages.ticketsViewAllLink)}
      </Link>
    </Block>
  )),
  ScreenLink: injectIntl(({ to, intl, label, blockTitle, ...props }) => (
    <Block
      title={intl.formatMessage({ id: blockTitle, defaultMessage: blockTitle })}
    >
      <Link title={props.description || ''} to={`/screens/${to}`} className="dpmsg-Button Button-FullWidth Button--primary">
        {intl.formatMessage({ id: label, defaultMessage: label })}
      </Link>
    </Block>
  )),
  ButtonLink: injectIntl(({ to, intl, label, blockTitle, description, ...props }) => (
    <Block title={intl.formatMessage({ id: blockTitle, defaultMessage: blockTitle })}>
      {description &&
        <div className="dpmsg-BlockText">
          {description}
        </div>
      }
      <Button  title={props.description || ''} to={`/screens/${to}`} width="full" color="primary">
        {intl.formatMessage({ id: label, defaultMessage: label })}
      </Button>
    </Block>
  ))
};

const Blocks = ({ blocks, agentsAvailable, activeChat, chatAvailable }) => (
  <ScrollArea
    stopScrollPropagation={true}
    horizontal={false}
    style={{ height: '100%' }}
  >
    {blocks.sort((blockA, blockB) => blockA.order - blockB.order).map(({ blockType, ...props }, index) => {
      if(blockType === 'StartChatBlock') {
        if (!Object.keys(agentsAvailable).length || !chatAvailable) {
          return null;
        }
        if (activeChat) {
          props.to = `active-chat/${activeChat}`;
          props.activeChat = activeChat;
        }
        props.agentsAvailable = agentsAvailable;
      }
      const Component = blocksMapping[blockType];
      return Component ? (
        <Component key={blockType + index} {...props} />
      ) : null;
    })}
    <Footer />
  </ScrollArea>
);

const mapStateToProps = (state) => ({
  agentsAvailable: getAgentsAvailable(state),
  activeChat:      getActiveChat(state),
  chatAvailable:   canUseChat(state)
});

export default connect(mapStateToProps)(Blocks);
