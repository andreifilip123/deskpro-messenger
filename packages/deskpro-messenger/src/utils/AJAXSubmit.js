// eslint-disable-next-line func-names
const AJAXSubmit = (function () {
  function submitData(oData, config) {
    /* the AJAX request... */
    const oAjaxReq = new XMLHttpRequest();
    oAjaxReq.submittedData = oData;

    if (config.updateProgress) {
      oAjaxReq.addEventListener('progress', config.updateProgress);
    }
    if (config.transferComplete) {
      oAjaxReq.addEventListener('load', config.transferComplete);
    }
    if (config.transferFailed) {
      oAjaxReq.addEventListener('error', config.transferFailed);
    }
    if (config.transferCanceled) {
      oAjaxReq.addEventListener('abort', config.transferCanceled);
    }

    oAjaxReq.onreadystatechange = function() { // Call a function when the state changes.
      if (this.readyState === XMLHttpRequest.DONE) {
        if (this.status >= 400) {
          let message = this.statusText;
          if (this.response && this.response.errors && this.response.errors.file) {
            message = this.response.errors.file;
          }
          config.transferFailed({
            code: this.status,
            message
          });
        }
      }
    }

    oAjaxReq.withCredentials = true;
    /* method is POST */
    oAjaxReq.responseType = 'json';
    oAjaxReq.open('post', oData.receiver, true);
    /* enctype is multipart/form-data */
    const sBoundary = `---------------------------${Date.now().toString(16)}`;
    oAjaxReq.setRequestHeader('Content-Type', `multipart/form-data; boundary=${sBoundary}`);
    if (config.requestHeaders) {
      for (const [name, value] of Object.entries(config.requestHeaders)) {
        oAjaxReq.setRequestHeader(name, value);
      }
    }
    oAjaxReq.send(`--${sBoundary}\r\n${
      oData.segments.join(`--${sBoundary}\r\n`)}--${sBoundary}--\r\n`);
  }

  function processStatus(oData, config = {}) {
    if (oData.status > 0) { return; }
    /* the form is now totally serialized! do something before sending it to the server... */
    // /* doSomething(oData); */
    // /* console.log("AJAXSubmit - The form is now serialized. Submitting..."); */
    submitData(oData, config);
  }

  function pushSegment(oFREvt) {
    this.owner.segments[this.segmentIdx] += `${oFREvt.target.result}\r\n`;
    this.owner.status = this.owner.status - 1;
    processStatus(this.owner, this.config);
  }

  function SubmitRequest(config) {
    this.receiver = config.url;
    this.status = 0;
    this.segments = [];
    for (let nFile = 0; nFile < config.files.length; nFile++) {
      const oFile = config.files[nFile];
      const oSegmReq = new FileReader();
      /* (custom properties:) */
      oSegmReq.segmentIdx = this.segments.length;
      oSegmReq.owner = this;
      oSegmReq.config = config;
      /* (end of custom properties) */
      oSegmReq.onload = pushSegment;
      this.segments.push(`Content-Disposition: form-data; name="file"; filename="${oFile.name}"\r\nContent-Type: ${oFile.type}\r\n\r\n`);
      this.status = this.status + 1;
      oSegmReq.readAsBinaryString(oFile);
    }
    this.segments.push(`Content-Disposition: form-data;`);
    processStatus(this, config);
  }

  // eslint-disable-next-line func-names
  return function (config) {
    if (!config.url) { return; }
    new SubmitRequest(config);
  };
}());

export default AJAXSubmit;
