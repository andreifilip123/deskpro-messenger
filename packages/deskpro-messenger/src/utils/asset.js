window.parent.DESKPRO_MESSENGER_OPTIONS =
  window.parent.DESKPRO_MESSENGER_OPTIONS || {};

export default (path) => {
  let edited = '';
  if (window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl) {
    edited = window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl.replace(/\/$/, "");
  }
  return `${edited}/assets/${path}`;
};
