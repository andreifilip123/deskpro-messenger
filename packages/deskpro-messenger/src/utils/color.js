/**
 * Returns true if opposite color is light or not.
 * @param {string} color Color hash.
 * @link https://stackoverflow.com/a/35970186
 */
export const isLightColor = (color) => {
  let hex = color.startsWith('#') ? color.slice(1) : color;
  // convert 3-digit hex to 6-digits.
  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  if (hex.length !== 6) {
    throw new Error('Invalid HEX color.');
  }
  const r = parseInt(hex.slice(0, 2), 16),
    g = parseInt(hex.slice(2, 4), 16),
    b = parseInt(hex.slice(4, 6), 16);
  // http://stackoverflow.com/a/3943023/112731
  return r * 0.299 + g * 0.587 + b * 0.114 > 186;
};
