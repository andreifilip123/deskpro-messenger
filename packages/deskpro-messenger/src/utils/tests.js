import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';

const customRender = (node, options) => {
  return render(<IntlProvider locale="en-US">{node}</IntlProvider>, options);
};

// re-export everything
export * from 'react-testing-library';

// override render method
export { customRender as render };
