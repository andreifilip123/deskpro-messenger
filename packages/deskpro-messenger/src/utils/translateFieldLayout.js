/**
 * Translates all strings (labels) in a field layout config.
 *
 * @param {Array} config Field Layout config
 * @param {function} formatMessage Callback from react-intl to translate a string.
 *
 * @returns new translated field layout config.
 */
export default (config, formatMessage) => {
  const t = (str) => formatMessage({ id: str, defaultMessage: str });

  return config.map((layout) => ({
    ...layout,
    fields: layout.fields.map((field) => {
      const newField = { ...field };
      newField.label = t(field.label);
      if (newField.placeholder) {
        newField.placeholder = t(field.placeholder);
      }
      if (Array.isArray(newField.validation)) {
        newField.validation = field.validation.map(
          (rule) =>
            typeof rule === 'object'
              ? { ...rule, message: t(rule.message) }
              : rule
        );
      }
      if (Array.isArray(newField.options)) {
        newField.options = field.options.map((option) => ({
          ...option,
          label: t(option.label)
        }));
      }
      if (
        typeof newField.dataSource === 'object' &&
        Array.isArray(newField.dataSource.getOptions)
      ) {
        newField.dataSource.getOptions = field.dataSource.getOptions.map(
          (option) => ({
            ...option,
            label: t(option.label)
          })
        );
      }
      return newField;
    })
  }));
};
