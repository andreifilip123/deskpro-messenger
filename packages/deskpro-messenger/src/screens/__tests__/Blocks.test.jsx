import React from 'react';
import { render } from '../../utils/tests';

import Blocks from '../Blocks';

const blocks = [
  {
    blockType: 'StartChatBlock',
    linkText: 'Start new chat',
    to: 'startChat'
  },
  {
    blockType: 'TicketsBlock'
  },
  {
    blockType: 'ScreenLink',
    blockTitle: 'tickets.create_form_block.header',
    label: 'tickets.create_form_block.link_label',
    to: 'newTicket'
  }
];

describe('<Blocks />', () => {
  it('should render correctly', () => {
    const { getByText, asFragment } = render(<Blocks blocks={blocks} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('Conversations')).toBeInTheDocument();
    expect(getByText('Start new chat')).toBeInTheDocument();
    expect(getByText('Your Tickets')).toBeInTheDocument();
    expect(getByText('tickets.create_form_block.header')).toBeInTheDocument();
    expect(
      getByText('tickets.create_form_block.link_label')
    ).toBeInTheDocument();
  });
});
