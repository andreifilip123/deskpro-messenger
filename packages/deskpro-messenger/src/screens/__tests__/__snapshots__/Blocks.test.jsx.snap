// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blocks /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="dpmsg-Block"
  >
    <div
      class="dpmsg-BlockWrapper"
    >
      <div
        class="dpmsg-BlockHeader"
      >
        Conversations
      </div>
      <a
        class="dpmsg-Button Button-FullWidth Button--primary"
        href="/screens/startChat"
      >
        Start new chat
      </a>
    </div>
  </div>
  <div
    class="dpmsg-Block"
  >
    <div
      class="dpmsg-BlockWrapper"
    >
      <div
        class="dpmsg-BlockHeader"
      >
        Your Tickets
      </div>
      <a
        href="/screens/tickets"
      >
        view all
      </a>
    </div>
  </div>
  <div
    class="dpmsg-Block"
  >
    <div
      class="dpmsg-BlockWrapper"
    >
      <div
        class="dpmsg-BlockHeader"
      >
        tickets.create_form_block.header
      </div>
      <a
        href="/screens/newTicket"
      >
        tickets.create_form_block.link_label
      </a>
    </div>
  </div>
</DocumentFragment>
`;
